src/Domain/UseCase/GetUserFeatures/GetUserFeatures.ts

```diff
@@ -0,0 +1,66 @@
+import { UseCaseInterface } from '../UseCaseInterface'
+import { inject, injectable } from 'inversify'
+import TYPES from '../../../Bootstrap/Types'
+import { GetUserFeaturesDto } from './GetUserFeaturesDto'
+import { UserRepositoryInterface } from '../../User/UserRepositoryInterface'
+import { GetUserFeaturesResponse } from './GetUserFeaturesResponse'
+import { RoleName, SubscriptionName } from '@standardnotes/auth'
+import { Feature } from '@standardnotes/features'
+import { UserSubscription } from '../../User/UserSubscription'
+import { Features } from '@standardnotes/features'
+
+
+@injectable()
+export class GetUserFeatures implements UseCaseInterface {
+  constructor(
+    @inject(TYPES.UserRepository) private userRepository: UserRepositoryInterface,
+  ) {
+  }
+
+  async execute(dto: GetUserFeaturesDto): Promise<GetUserFeaturesResponse> {
+    const { userUuid } = dto
+
+    const user = await this.userRepository.findOneByUuid(userUuid)
+
+    if (user === undefined) {
+      return {
+        success: false,
+        error: {
+          message: `User ${userUuid} not found.`,
+        },
+      }
+    }
+
+    const userRoles = await user.roles
+    const userSubscriptions = await user.subscriptions
+
+    const features = await Promise.all(userRoles.map(async (role) => {
+      const subscriptionName = this.roleNameToSubscriptionNameMap.get(role.name as RoleName)
+      const expiresAt = (userSubscriptions.find(subscription => subscription.planName === subscriptionName) as UserSubscription).endsAt
+
+      const permissions = await role.permissions
+      const featuresWithExpirationDate = permissions.map(permission => {
+        const featureItem = Features.find(feature => feature.identifier === permission.name)
+
+        return {
+          ...featureItem,
+          expiresAt,
+        }
+      })
+
+      return featuresWithExpirationDate
+    }))
+
+    return {
+      success: true,
+      userUuid,
+      features: features.flat() as Feature[],
```

Comment:
Perhaps we should make this a Set so that there aren't any duplicate features?

Reply:
Good point, it raises another question though: Let's say `Permission 1` belongs to `Role 1` and `Role 2`, expiration date of `Role 1` is September 2021 and for `Role 2` is March 2022. In such case I think we should set `expiresAt` for the feature that matches `Permission 1` to March 2022, is that correct?

Reply:
@vardan-arm correct.

TL;DR: longest subscription(role) that entitles you to a feature is the dominant one.

so the `Permission X` that has the longest expiration date should overwrite all `Permission X'` in the set that are it's duplicates with lesser expiration date.